File: backend/app/routers/facilities.py
"""FastAPI router for facility endpoints.

Routes:
  GET /api/v1/facilities
  GET /api/v1/facilities/{tri_facility_id}
  GET /api/v1/facilities/{tri_facility_id}/releases

Phase 2 — stories 2.1.1, 2.1.2, 2.2.x, 2.3.x.
"""

from __future__ import annotations

import datetime
import logging
from typing import Annotated, Any, Literal

from fastapi import APIRouter, Depends, HTTPException, Query
from fastapi.exceptions import RequestValidationError
from sqlalchemy.ext.asyncio import AsyncSession

from app.database import get_db
from app.schemas.facility import (
    FacilityCollection,
    FacilityDetail,
    ReleaseEventSchema,
    SiteSearchResult,
)
from app.services.facility_service import (
    get_all_facilities_browse,
    get_facilities_near,
    get_facility_detail,
    search_facilities,
)
from app.services.release_service import get_facility_releases

logger = logging.getLogger(__name__)

router = APIRouter(tags=["facilities"])

_ReleaseMedia = Literal["air", "water", "land", "underground"]


@router.get("/facilities/browse", response_model=FacilityCollection)
async def browse_all_facilities(
    year: Annotated[
        int | None,
        Query(description="Reporting year (omit for latest available)"),
    ] = None,
    chemical: Annotated[
        str | None,
        Query(description="Filter by chemical name (partial, case-insensitive)"),
    ] = None,
    medium: Annotated[
        _ReleaseMedia | None,
        Query(description="Release medium: air | water | land | underground"),
    ] = None,
    state: Annotated[
        str | None,
        Query(max_length=2, description="2-letter state code"),
    ] = None,
    bbox: Annotated[
        str | None,
        Query(description="Bounding box: min_lon,min_lat,max_lon,max_lat"),
    ] = None,
    exact_match: Annotated[
        bool,
        Query(description="Skip chemical family expansion (search exact term only)"),
    ] = False,
    limit: Annotated[int, Query(ge=1, le=50000)] = 50000,
    db: AsyncSession = Depends(get_db),
) -> FacilityCollection:
    """Browse mode: fetch ALL TRI facilities without radius constraint.

    Used for the initial map view showing all facilities nationwide.
    Optional bbox parameter filters to viewport bounds.
    """
    return await get_all_facilities_browse(
        session=db,
        year=year,
        chemical=chemical,
        medium=medium,
        state=state,
        bbox=bbox,
        exact_match=exact_match,
        limit=limit,
    )


@router.get("/facilities/search", response_model=list[SiteSearchResult])
async def search_facilities_endpoint(
    q: Annotated[
        str,
        Query(
            min_length=2,
            max_length=100,
            description="Search query (TRI ID, EPA ID, or site name; min 2 chars)",
        ),
    ],
    state: Annotated[
        str | None,
        Query(max_length=2, description="Filter to 2-letter state code"),
    ] = None,
    limit: Annotated[int, Query(ge=1, le=50)] = 10,
    db: AsyncSession = Depends(get_db),
) -> list[SiteSearchResult]:
    """Search TRI facilities and Superfund sites by ID or name (ADR-010).

    Searches both TRI facilities (by TRI ID or name) and Superfund sites
    (by EPA ID or name). Results are merged and ranked by relevance score:
    - Exact TRI ID or EPA ID match: score 1.0
    - ID prefix: score 0.95
    - Exact name match: score 0.90
    - Name prefix: score 0.80
    - Name contains: score 0.60
    - ID contains: score 0.50

    Returns empty array (not 404) when no sites match.
    Results ordered by relevance_score DESC, then name ASC.
    Use site_type field to distinguish TRI ("tri") from Superfund ("superfund").
    """
    return await search_facilities(db, q, state, limit)


@router.get("/facilities", response_model=FacilityCollection)
async def list_facilities(
    lat: Annotated[
        float,
        Query(ge=-90.0, le=90.0, description="Center latitude (WGS84)"),
    ],
    lon: Annotated[
        float,
        Query(ge=-180.0, le=180.0, description="Center longitude (WGS84)"),
    ],
    radius_miles: Annotated[
        float,
        Query(gt=0, le=500.0, description="Search radius in miles (max 500)"),
    ],
    bbox: Annotated[
        str | None,
        Query(description="Bounding box: minLon,minLat,maxLon,maxLat"),
    ] = None,
    year: Annotated[
        int | None,
        Query(description="Reporting year (omit for latest available)"),
    ] = None,
    chemical: Annotated[
        str | None,
        Query(description="Filter by chemical name (partial, case-insensitive)"),
    ] = None,
    naics: Annotated[
        str | None,
        Query(description="Filter by NAICS code prefix"),
    ] = None,
    medium: Annotated[
        _ReleaseMedia | None,
        Query(description="Release medium: air | water | land | underground"),
    ] = None,
    state: Annotated[
        str | None,
        Query(max_length=2, description="2-letter state code"),
    ] = None,
    restrict_to_state: Annotated[bool, Query()] = False,
    exact_match: Annotated[
        bool,
        Query(description="Skip chemical family expansion (search exact term only)"),
    ] = False,
    limit: Annotated[int, Query(ge=1, le=2000)] = 500,
    db: AsyncSession = Depends(get_db),
) -> FacilityCollection:
    if restrict_to_state and (not state or len(state) != 2):
        raise RequestValidationError(
            [
                {
                    "type": "value_error",
                    "loc": ("query", "state"),
                    "msg": "restrict_to_state=true requires a 2-character state code",
                    "input": state,
                }
            ]
        )

    raw_query: dict[str, Any] = {
        "lat": lat,
        "lon": lon,
        "radius_miles": radius_miles,
        "year": year,
        "chemical": chemical,
        "naics": naics,
        "medium": medium,
        "state": state,
        "restrict_to_state": restrict_to_state,
    }
    return await get_facilities_near(
        session=db,
        lat=lat,
        lon=lon,
        radius_miles=radius_miles,
        bbox=bbox,
        year=year,
        chemical=chemical,
        naics=naics,
        medium=medium,
        state=state,
        restrict_to_state=restrict_to_state,
        exact_match=exact_match,
        limit=limit,
        raw_query=raw_query,
    )


@router.get(
    "/facilities/{tri_facility_id}",
    response_model=FacilityDetail,
)
async def get_facility(
    tri_facility_id: str,
    year: Annotated[
        int | None,
        Query(description="Filter top chemicals and totals to this reporting year (omit for all years)"),
    ] = None,
    db: AsyncSession = Depends(get_db),
) -> FacilityDetail:
    detail = await get_facility_detail(db, tri_facility_id, year=year)
    if detail is None:
        raise HTTPException(status_code=404, detail="Facility not found")
    return detail


@router.get(
    "/facilities/{tri_facility_id}/releases",
    response_model=list[ReleaseEventSchema],
)
async def list_facility_releases(
    tri_facility_id: str,
    from_year: Annotated[int | None, Query()] = None,
    to_year: Annotated[int | None, Query()] = None,
    chemical_id: Annotated[int | None, Query()] = None,
    medium: Annotated[_ReleaseMedia | None, Query()] = None,
    db: AsyncSession = Depends(get_db),
) -> list[ReleaseEventSchema]:
    today = datetime.date.today()
    eff_from = from_year if from_year is not None else today.year - 14
    eff_to = to_year if to_year is not None else today.year

    releases = await get_facility_releases(
        session=db,
        tri_facility_id=tri_facility_id,
        from_year=eff_from,
        to_year=eff_to,
        chemical_id=chemical_id,
        medium=medium,
    )
    if releases is None:
        raise HTTPException(status_code=404, detail="Facility not found")
    return releases

File: frontend/src/api/facilities.ts
/**
 * Typed API client — Facilities (stories 2.1.x, 2.2.x, ADR-010).
 * All requests go through the Vite proxy (/api → http://backend:8000) in dev/Docker.
 * When VITE_API_BASE_URL is set (outside Docker), requests are direct.
 */
import type {
  FacilityCollection,
  FacilityDetail,
  FacilitySearchResult,
  ReleaseEvent,
  SearchParams,
} from './types'

const API_BASE = import.meta.env.VITE_API_BASE_URL ?? ''

/**
 * Builds the query URL for GET /api/v1/facilities.
 * Security: all params are passed via URLSearchParams — no string interpolation with user input.
 */
function buildFacilitySearchUrl(params: SearchParams): string {
  const p = new URLSearchParams({
    lat: String(params.lat),
    lon: String(params.lon),
    radius_miles: String(params.radiusMiles),
  })
  if (params.chemical) p.set('chemical', params.chemical)
  if (params.year) p.set('year', params.year)
  if (params.medium) p.set('medium', params.medium)
  if (params.state) p.set('state', params.state)
  if (params.restrictToState) p.set('restrict_to_state', 'true')
  if (params.bbox) p.set('bbox', params.bbox.join(','))
  if (params.exactMatch) p.set('exact_match', 'true')
  return `${API_BASE}/api/v1/facilities?${p.toString()}`
}

/** GET /api/v1/facilities — radius + filter search */
export async function fetchFacilities(
  params: SearchParams,
  signal?: AbortSignal,
): Promise<FacilityCollection> {
  const url = buildFacilitySearchUrl(params)
  const res = await fetch(url, signal ? { signal } : {})
  if (!res.ok) throw new Error(`Facility search failed: ${res.status}`)
  const data = await res.json() as FacilityCollection
  return data
}

/** GET /api/v1/facilities/{id} — single facility detail */
export async function fetchFacilityDetail(id: string): Promise<FacilityDetail> {
  const res = await fetch(`${API_BASE}/api/v1/facilities/${encodeURIComponent(id)}`)
  if (!res.ok) throw new Error(`Facility detail failed: ${res.status}`)
  return res.json() as Promise<FacilityDetail>
}

/** GET /api/v1/facilities/{id}/releases — 15-year time series */
export async function fetchFacilityReleases(
  id: string,
  fromYear?: number,
  toYear?: number,
): Promise<ReleaseEvent[]> {
  const p = new URLSearchParams()
  if (fromYear != null) p.set('from_year', String(fromYear))
  if (toYear != null) p.set('to_year', String(toYear))
  const qs = p.toString() ? `?${p.toString()}` : ''
  const res = await fetch(`${API_BASE}/api/v1/facilities/${encodeURIComponent(id)}/releases${qs}`)
  if (!res.ok) throw new Error(`Facility releases failed: ${res.status}`)
  return res.json() as Promise<ReleaseEvent[]>
}

export interface BrowseParams {
  year?: string
  chemical?: string
  medium?: string
  state?: string
  /** ADR-007: Skip chemical family expansion (search exact term only) */
  exactMatch?: boolean
}

/** GET /api/v1/facilities/browse — all facilities without radius constraint (browse mode) */
export async function fetchAllFacilitiesBrowse(
  params: BrowseParams = {},
  signal?: AbortSignal,
): Promise<FacilityCollection> {
  const p = new URLSearchParams()
  if (params.year) p.set('year', params.year)
  if (params.chemical) p.set('chemical', params.chemical)
  if (params.medium) p.set('medium', params.medium)
  if (params.state) p.set('state', params.state)
  if (params.exactMatch) p.set('exact_match', 'true')
  const qs = p.toString() ? `?${p.toString()}` : ''
  const url = `${API_BASE}/api/v1/facilities/browse${qs}`
  const res = await fetch(url, signal ? { signal } : {})
  if (!res.ok) throw new Error(`Browse facilities failed: ${res.status}`)
  const data = await res.json() as FacilityCollection
  return data
}

/**
 * GET /api/v1/facilities/search — autocomplete by TRI ID or name (ADR-010).
 *
 * Returns ranked results with:
 * - Exact TRI ID match: score 1.0
 * - TRI ID prefix: score 0.95
 * - Exact name match: score 0.90
 * - Name prefix: score 0.80
 * - Name contains: score 0.60
 * - TRI ID contains: score 0.50
 */
export async function searchFacilities(
  q: string,
  state?: string,
  limit = 10,
  signal?: AbortSignal,
): Promise<FacilitySearchResult[]> {
  const p = new URLSearchParams({ q })
  if (state) p.set('state', state)
  if (limit !== 10) p.set('limit', String(limit))
  const res = await fetch(
    `${API_BASE}/api/v1/facilities/search?${p.toString()}`,
    signal ? { signal } : {},
  )
  if (!res.ok) throw new Error(`Facility search failed: ${res.status}`)
  return res.json() as Promise<FacilitySearchResult[]>
}
